### PR TITLE
feat: refactor cross entropy, add CUDA, fix grad test

### PR DESCRIPTION
**Pull request by JohannesGaessler**
I'm currently working on enabling MNIST training for backends other than CPU. As part of that I'm adding CUDA support for cross entropy loss which I'm spinning out into a separate PR since I think it will make reviewing easier. This PR makes the following changes:

* Refactor the cross entropy CPU code to optimize out the logarithm in $\log(\exp(\mathrm{logit}_i - \mathrm{max}))$. This also removes the need for the workaround where the softmax is scaled from $[0,1]$ to $[\epsilon, 1]$, simplifying the code.
* Increase `eps` for the test in `tests/test-grad0` to reduce the impact of machine precision on the numerical gradient calculation. Increase the range of the logits to ensure that cross entropy is sufficiently linear on the scale of `eps`. On master cross entropy is failing for 123/1000 iterations, with the tuned parameters it's failing for 0/1000 iterations (both for the code on master and the code in this PR).
* Add a CUDA implementation for cross entropy loss. Expose `sum_rows_f32_cuda` so that the code can be re-used to combine the partial cross entropy results into a scalar value.

### CODE REVIEW

File: tests/test-backend-ops.cpp [hunk 1]
```diff
@@ -1730,6 +1730,27 @@ struct test_flash_attn_ext : public test_case {
     }
 };
 
+// GGML_OP_CROSS_ENTROPY_LOSS
+struct test_cross_entropy_loss : public test_case {
+    const ggml_type type;
+    const std::array<int64_t, 4> ne;
+
+    std::string vars() override {
+        return VARS_TO_STR1(ne);
```
Comment (slaren) [inline tests/test-backend-ops.cpp:1739]: This should also include the parameter `type`.

**Review by ggerganov — APPROVED** [review]

File: src/ggml-cuda/cross-entropy-loss.cu [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    GGML_ASSERT( dst->type == GGML_TYPE_F32);
+
+    GGML_ASSERT(ggml_is_contiguous(src0));
+    GGML_ASSERT(ggml_is_contiguous(src1));
+    GGML_ASSERT(ggml_is_contiguous(dst));
+
+    const int64_t ne00  = src0->ne[0];
+    const int64_t nrows = ggml_nrows(src0);
+
+    const int64_t ne01_padded = GGML_PAD(nrows, WARP_SIZE);
```
Comment (slaren) [inline src/ggml-cuda/cross-entropy-loss.cu:89]: This variable is unused, and generates a warning.

**Review by slaren — APPROVED** [review]
